src/openfermion/measurements/_qubit_partitioning.py

```diff
@@ -0,0 +1,167 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+
+""" Code to generate Pauli strings for measurement of local operators"""
+from __future__ import division
+import numpy
+from itertools import chain
+try:
+    from itertools import zip_longest
+except ImportError:
+    from itertools import izip_longest as zip_longest
+
+
+def binary_partition_iterator(qubit_list, num_iterations=None):
+    '''Generator for a list of 2-partitions of N qubits
+    such that all pairs of qubits are split in at least one partition,
+    This follows a variation on ArXiv:1908.0562 - instead of
+    explicitly partitioning the list based on the binary indices of
+    the qubits, we repeatedly divide the list in two and then
+    zip it back together.
+
+    Args:
+        qubit_list(list): list of qubits to be partitioned
+        num_iterations(int or None): number of iterations to perform.
+            If None, will be set to ceil(log2(len(qubit_list)))
+
+    Returns:
+        partition(iterator of lists of lists): the required partitioning
+    '''
+
+    # Some edge cases
+    if num_iterations is not None and num_iterations == 0:
+        return
+    num_qubits = len(qubit_list)
+    if num_qubits < 2:
+        raise ValueError('Need at least 2 qubits to partition')
+    if num_qubits == 2:
+        yield [[qubit_list[0]], [qubit_list[1]]]
+        return
+
+    num_iterations = num_iterations or\
+        int(numpy.ceil(numpy.log2(num_qubits)))
+
+    # Calculate the point where we need to split the list each time.
+    half_point = int(numpy.ceil(num_qubits/2))
+
+    # Repeat the division and zip steps as many times
+    # as required.
+    for j in range(num_iterations):
+        # Divide the qubit list in two and return it
+        partition = [qubit_list[:half_point],
+                     qubit_list[half_point:]]
```

Comment:
Yield a tuple instead of a list.

Reply:
Sounds good, I've done this throughout the three functions in the file and added some checks in the test suite to ensure that they are indeed returning tuples.